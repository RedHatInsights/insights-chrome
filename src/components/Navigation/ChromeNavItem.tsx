import React, { useEffect, useMemo } from 'react';
import { Icon } from '@patternfly/react-core/dist/dynamic/components/Icon';
import { NavItem } from '@patternfly/react-core/dist/dynamic/components/Nav';
import { Tooltip } from '@patternfly/react-core/dist/dynamic/components/Tooltip';
import ExternalLinkAltIcon from '@patternfly/react-icons/dist/dynamic/icons/external-link-alt-icon';
import FlaskIcon from '@patternfly/react-icons/dist/dynamic/icons/flask-icon';
import BellIcon from '@patternfly/react-icons/dist/dynamic/icons/bell-icon';
import StarIcon from '@patternfly/react-icons/dist/dynamic/icons/star-icon';
import { titleCase } from 'title-case';
import classNames from 'classnames';
import get from 'lodash/get';
import { useAtomValue, useSetAtom } from 'jotai';

import ChromeLink, { LinkWrapperProps } from '../ChromeLink/ChromeLink';
import { useSelector } from 'react-redux';
import { ChromeNavItemProps } from '../../@types/types';
import useFavoritePagesWrapper from '../../hooks/useFavoritePagesWrapper';
import { isPreviewAtom } from '../../state/atoms/releaseAtom';
import { activeProductAtom } from '../../state/atoms/activeProductAtom';

const ChromeNavItem = ({
  appId,
  className,
  href,
  isHidden,
  ignoreCase,
  title,
  isExternal,
  isBeta: isBetaEnv,
  active,
  product,
  notifier = '',
}: ChromeNavItemProps) => {
  const isPreview = useAtomValue(isPreviewAtom);
  const hasNotifier = useSelector((state) => get(state, notifier));
  const markActiveProduct = useSetAtom(activeProductAtom);
  const { favoritePages } = useFavoritePagesWrapper();
  const isFavorited = useMemo(() => favoritePages.find(({ favorite, pathname }) => favorite && pathname === href), [href, favoritePages]);

  useEffect(() => {
    if (active) {
      markActiveProduct(product);
    }
  }, [active]);

  if (isHidden) {
    return null;
  }

  return (
    <NavItem
      className={classNames('chr-c-navigation__item', className, { 'chr-c-navigation__with-notifier': hasNotifier })}
      itemID={href}
      data-quickstart-id={href}
      preventDefault
      isActive={active}
      to={href}
      ouiaId={title}
      component={(props: LinkWrapperProps) => <ChromeLink {...props} isBeta={isBetaEnv} isExternal={isExternal} appId={appId} />}
    >
      {typeof title === 'string' && !ignoreCase ? titleCase(title) : title}{' '}
      {isExternal && (
        <Icon isInline>
          <ExternalLinkAltIcon />
        </Icon>
      )}
      {isBetaEnv && !isPreview && !isExternal && (
        <Tooltip position={'right'} content={<div>This service is a Preview.</div>}>
          <Icon className="chr-c-navigation__beta-icon" isInline>
            <FlaskIcon />
          </Icon>
        </Tooltip>
      )}
      {isFavorited && (
        <Icon>
          <StarIcon color="var(--pf-t--temp--dev--tbd)"/* CODEMODS: original v5 color was --pf-v5-global--palette--gold-400 */ />
        </Icon>
      )}
      {hasNotifier && (
        <Icon size="md">
          <BellIcon className="notifier-icon" color="var(--pf-v5-global--default-color--200)" />
        </Icon>
      )}
    </NavItem>
  );
};

export default ChromeNavItem;
