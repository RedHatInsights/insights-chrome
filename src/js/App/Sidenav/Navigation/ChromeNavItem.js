import React from 'react';
import PropTypes from 'prop-types';
import { NavItem } from '@patternfly/react-core';
import ExternalLinkAltIcon from '@patternfly/react-icons/dist/js/icons/external-link-alt-icon';
import { titleCase } from 'title-case';
import classNames from 'classnames';

import { isBeta } from '../../../utils';
import { betaBadge } from '../../Header/Tools';
import ChromeLink from './ChromeLink';

const ChromeNavItem = ({ appId, className, href, isHidden, ignoreCase, title, isExternal, isBeta: isBetaEnv, active }) => {
  if (isHidden) {
    return null;
  }

  return (
    <NavItem
      className={classNames(className, { 'ins-c-navigation__additional-links': isExternal })}
      itemID={href}
      data-quickstart-id={href}
      preventDefault
      isActive={active}
      to={href}
      ouiaId={title}
      component={(props) => <ChromeLink {...props} isBeta={isBetaEnv} isExternal={isExternal} appId={appId} />}
    >
      {typeof title === 'string' && !ignoreCase ? titleCase(title) : title} {isExternal && <ExternalLinkAltIcon />}
      {isBetaEnv && !isBeta() && !isExternal && betaBadge('ins-c-navigation__beta-badge')}
    </NavItem>
  );
};

ChromeNavItem.propTypes = {
  isHidden: PropTypes.bool,
  ignoreCase: PropTypes.bool,
  title: PropTypes.node,
  isExternal: PropTypes.bool,
  isBeta: PropTypes.bool,
  href: PropTypes.string.isRequired,
  className: PropTypes.string,
  active: PropTypes.bool,
  appId: PropTypes.string,
};

export default ChromeNavItem;
