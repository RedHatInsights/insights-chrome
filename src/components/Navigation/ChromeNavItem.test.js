/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import createMockStore from 'redux-mock-store';
import NavContext from './navContext';
import ChromeNavItem from './ChromeNavItem';

const NavContextWrapper = ({
  store,
  providerValue = {
    onLinkClick: jest.fn(),
  },
  children,
  initialEntries,
}) => (
  <MemoryRouter initialEntries={initialEntries}>
    <Provider store={store}>
      <NavContext.Provider value={providerValue}>{children}</NavContext.Provider>
    </Provider>
  </MemoryRouter>
);

describe('ChromeNavItem', () => {
  const mockStore = createMockStore();
  const linkTitle = 'Foo';
  const testProps = {
    appId: 'testModule',
    href: '/foo',
    title: linkTitle,
  };
  const store = mockStore({
    chrome: {
      activeModule: 'testModule',
      moduleRoutes: [],
      modules: {
        testModule: {},
      },
    },
  });

  test('should not render nav item', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItem isHidden {...testProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(linkTitle)).toHaveLength(0);
  });

  test('should render nav item', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItem {...testProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText('beta')).toHaveLength(0);
    expect(queryAllByText(linkTitle)).toHaveLength(1);
  });

  test('should render nav item with an external icon', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItem {...testProps} isExternal />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText('beta')).toHaveLength(0);
    expect(queryAllByText(linkTitle)).toHaveLength(1);
  });

  test('should render nav item with a node title', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItem {...testProps} title={<span data-testid="test-span">Node title</span>} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText('beta')).toHaveLength(0);
    expect(queryAllByText('Node title')).toHaveLength(1);
  });

  test('should render nav item with beta badge', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItem {...testProps} isBeta />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText('beta')).toHaveLength(1);
  });
});
