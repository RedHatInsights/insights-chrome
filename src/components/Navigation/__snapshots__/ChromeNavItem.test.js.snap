// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeNavItem should not render nav item 1`] = `<div />`;

exports[`ChromeNavItem should render nav item 1`] = `
<div>
  <li
    class="pf-c-nav__item"
    data-ouia-component-id="Foo"
    data-ouia-component-type="PF4/NavItem"
    data-ouia-safe="true"
  >
    <a
      class="pf-c-nav__link"
      data-quickstart-id="foo"
      data-testid="router-link"
      href="/foo"
    >
      Foo
       
    </a>
  </li>
</div>
`;

exports[`ChromeNavItem should render nav item with a node title 1`] = `
<div>
  <li
    class="pf-c-nav__item"
    data-ouia-component-id="[object Object]"
    data-ouia-component-type="PF4/NavItem"
    data-ouia-safe="true"
  >
    <a
      class="pf-c-nav__link"
      data-quickstart-id="foo"
      data-testid="router-link"
      href="/foo"
    >
      <span
        data-testid="test-span"
      >
        Node title
      </span>
       
    </a>
  </li>
</div>
`;

exports[`ChromeNavItem should render nav item with an external icon 1`] = `
<div>
  <li
    class="pf-c-nav__item chr-c-navigation__additional-links"
    data-ouia-component-id="Foo"
    data-ouia-component-type="PF4/NavItem"
    data-ouia-safe="true"
  >
    <a
      class="pf-c-nav__link chr-c-navigation__additional-links"
      data-quickstart-id="/foo"
      data-testid="native-link"
      href="/foo"
      itemid="/foo"
      rel="noreferrer noopener"
      target="_blank"
    >
      Foo
       
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
        />
      </svg>
    </a>
  </li>
</div>
`;

exports[`ChromeNavItem should render nav item with beta badge 1`] = `
<div>
  <li
    class="pf-c-nav__item"
    data-ouia-component-id="Foo"
    data-ouia-component-type="PF4/NavItem"
    data-ouia-safe="true"
  >
    <a
      class="pf-c-nav__link"
      data-quickstart-id="foo"
      data-testid="router-link"
      href="/foo"
    >
      Foo
       
      <span
        class="pf-c-badge pf-m-unread chr-c-toolbar__beta-badge chr-c-navigation__beta-badge"
      >
        beta
      </span>
    </a>
  </li>
</div>
`;
